app/controllers/challenge.py

```diff
@@ -1,35 +1,87 @@
 import os
 from pathlib import Path
-from flask import render_template, request, redirect, url_for
-
-from app import app
+from sqlalchemy.exc import IntegrityError
+from flask import render_template, request, redirect, url_for, abort
+from flask_login import current_user
+from functools import wraps
+
+from app import app, login_manager, db
+from app.models.challenge import Challenge
+from app.models.user_solution import create_user_solution
+
+
+def challenge_access_required(func):
+    """
+    Decorator for checking if a user has access to a specific challenge
+
+    This decorator expects that the challenge title is passed in a keyword argument
+    called `name`
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if not current_user.is_authenticated:
+            return login_manager.unauthorized()
+        challenge_title = kwargs['name']
+        if not challenge_title:
+            # this should only happen when the controller has not been setup
+            # correctly
+            return abort(500)
+
+        # Get the order number for the last challenge the user solved
+        latest_user_solution = current_user.latest_solution
+        if not latest_user_solution:
+            challenge_accessed = Challenge.query.filter(
+                (Challenge.order_num == 0) & (Challenge.title == challenge_title)
+            ).first()
+        else:
+            challenge_num = latest_user_solution.challenge.order_num
+            challenge_accessed = Challenge.query.filter(
+                (Challenge.order_num <= challenge_num + 1) & (Challenge.title == challenge_title)
+            ).first()
+
+        if not challenge_accessed:
+            # This will be returned if the queries did not return any Challenges
+            return login_manager.unauthorized()
```

Comment:
I think this should be abort(403). Now it gives 401, which means the user needs to log in.

Reply:
Made an issue (#45)